Implement enough of geckolib to cascade the first node

bholley opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10697)

<!-- Reviewable:end -->

bholley committed:
Manually implement the display property for now.

bholley committed:
Track cloning separately from additional_methods and auto-generate clone methods.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties.mako.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

bholley commented:
r? @SimonSapin

bholley committed:
Use strings rather than arrays for skip_longhands and skip_additionals.

This matches what we do in components/style.

bholley committed:
Get rid of some of our custom additional methods in favor of clonable keywords.

Now that I'm pretty convinced we can auto-generate these and make them fast, we can
ditch the additional complexity.

bholley committed:
Make predefined_type use kwargs.

bholley commented:
Added some additional stuff from today.

bholley committed:
Manually implement overflow-y.

bholley committed:
Regenerate bindings with gfx/2d/Types.h included.

bholley committed:
Add support for border-*-style.

bholley committed:
Handle BORDER_STYLE_MASK.

Gecko uses the higher bits of certain border-style-like properties to store other things.
This duplicates the logic in Gecko's {Get,Set}BorderStyle and similar.

bholley committed:
Support the text-decoration-style property.

bholley committed:
Implement border_SIDE_has_nonzero_width.

bholley committed:
Implement outline-style and outline_has_nonzero_width.

bholley committed:
Implement font-size.

bholley committed:
Implement border-*-width.

bholley committed:
Unstub text-orientation.

The spec has changed a bit since the servo implementation. We still need to alias
sideways-right to sideways and add other writing modes, but we can do that later.

bholley committed:
Stub out is_multicol for now.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10749) made this pull request unmergeable. Please resolve the merge conflicts.

bholley committed:
Regenerate style struct bindings with latest bindgen and gecko.

bholley committed:
Remove no_move annotation now that gecko style structs are movable.

bholley committed:
Replace panics with warnings for unimplemented property setters and copiers.

bholley commented:
r? @SimonSapin

Renaming to reflect the fact that this is a catch-all for the geckolib work I did this week.

With this PR, we make it through the first cascaded node and panic in Servo_GetComputedValuesForAnonymousBox. @emilio is going to work on that next.

CC @heycam

bholley commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 735f1ff with merge 929aaa3...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/502)

bholley commented:
@bors-servo retry

On phone.

bors-servo commented:
:hourglass: Trying commit 735f1ff with merge e125a2b...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/504)

SimonSapin commented:
Style crate changes look good.

---

Reviewed 3 of 5 files at r8, 1 of 3 files at r10, 1 of 2 files at r14, 1 of 2 files at r20.
Review status: 6 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10697#-:-KFxEcRLQ_v_amg3VT7x:-1660219074)*
<!-- Sent from Reviewable.io -->

emilio commented on ports/geckolib/properties.mako.rs:
```diff
@@ -118,16 +118,35 @@ pub struct ${style_struct.gecko_struct_name};
+def get_gecko_property(ffi_name):
+    if is_border_style_masked(ffi_name):
+        return "(self.gecko.%s & (gecko_style_structs::BORDER_STYLE_MASK as u8))" % ffi_name
+    else:
```
nit: unneeded return?

emilio commented on ports/geckolib/properties.mako.rs:
```diff
@@ -118,16 +118,35 @@ pub struct ${style_struct.gecko_struct_name};
+    if is_border_style_masked(ffi_name):
+        return "self.gecko.%s &= !(gecko_style_structs::BORDER_STYLE_MASK as u8);" % ffi_name + \
+               "self.gecko.%s |= %s as u8;" % (ffi_name, expr)
+    else:
```
ditto

bholley commented:
@emilio the return statements are necessary, because it's python code, and a regular def rather than a mako def. I used a regular def here to do inline expressions, since mako defs insert newlines before and after.

bholley commented:
@bors-servo r=SimonSapin

bors-servo commented:
:pushpin: Commit 735f1ff has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 735f1ff523d6caa2baaaa1125c8db06e894092dc -->

jdm commented:
@bors-servo: try- retry clean

bors-servo commented:
:hourglass: Testing commit 735f1ff with merge c372dee...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/469), [arm32](http://build.servo.org/builders/arm32/builds/501), [arm64](http://build.servo.org/builders/arm64/builds/486), [linux-dev](http://build.servo.org/builders/linux-dev/builds/496), [linux-rel](http://build.servo.org/builders/linux-rel/builds/546), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/510), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/496), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/523)
